Drive Train Turn pid and Pigeon Sensor

Lockscout opened the pull request:
The PID turning command is fully functional for now, unless people come up with things to add in the future. I also added a Pigeon sensor so that it doesn't need re declared every single time we use it.

You supply it the degrees to turn, (negative for left positive for right), and it will turn that many degrees. The pid will need to be re tuned when we get the new robot.

Maps to Kanban issue #11

Lockscout committed:
Merge branch 'master' into DriveTrainPID

Lockscout committed:
Implemented the DrivetrainPIDturning command and pigeon sensor

Added the pigeon sensor and the DrivetrainPIDturning command. Also made the talons public. thrown together tank drive code added as well. Will fully implement when drivebase is done.

Commit:
Merge branch 'master' into DriveTrainPID

nrgill28 commented on src/main/java/frc/robot/commands/DrivetrainPIDTurnDelta.java:
```diff
@@ -0,0 +1,75 @@
+
+    //Delcare PD variables
+    private double kPgain = 0.01;
+    private double kDgain = 0.0016;
```
These are a great use for FluidConstants!

nrgill28 commented on .idea/modules/2020-2706-Robot-Code.main.iml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Is this file necessary?

nrgill28 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -56,7 +59,7 @@ private void configureButtonBindings() {
     DriveBase.getInstance().setDefaultCommand(driveCommand);
     
-  
+    new JoystickButton(driverStick, XboxValue.XBOX_A_BUTTON.getPort()).whenHeld(new DrivetrainPIDTurnDelta(90));
```
Remove controller bindings used for debug!

nrgill28 commented on src/main/java/frc/robot/commands/DrivetrainPIDTurnDelta.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    @Override
+    public void end(boolean interrupted) {
```
Remove any methods that you leave unimplemented

Lockscout committed:
Fixed pull request issues, updated fluid constants a bit

Fixed the issues brought up in the pull request. Nathan fixed pull request stuff for fluid constants.

nrgill28 reviewed (approved):
Looks good. For anyone wondering, the changes to the FluidConstant class were made to prevent code which only gets the values from FluidConstants from breaking in case the FluidConstant class changes. (Because we might change how FluidConstant works later, but as long as we still implement `Supplier<T>` no code should break!)

georgetzavelas commented on src/main/java/frc/robot/subsystems/DriveBase.java:
```diff
@@ -32,7 +32,7 @@
 
     // The drivebase talons
-    private WPI_TalonSRX leftFrontTalon, leftRearTalon, rightFrontTalon, rightRearTalon;
+    public WPI_TalonSRX leftFrontTalon, leftRearTalon, rightFrontTalon, rightRearTalon;
```
Curious why you are making these public instead of leaving them private?

kevlam2706 replied:
Ah yes. Because he needed to access the talon object in order to pass it to the Pigeon constructor, since that is how you tell the Pigeon that it is attached to a Talon.

My question is - if the Pigeon and the Talon are so strongly interrelated -- AND the Pigeon is physically part of the drivebase -- should the Pigeon be a member of DriveBase?

kevlam2706 replied:
agreed, see my previous comments about rearchitecting where to put the pigeon. If it needs to know about the Talon (which it does) then maybe it should be managed here along with the Talon.

georgetzavelas commented on .idea/misc.xml:
```diff
@@ -1,5 +1,5 @@
 <project version="4">
   <component name="ExternalStorageConfigurationManager" enabled="true" />
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_11" default="false" project-jdk-name="11.0.5" project-jdk-type="JavaSDK" />
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_11" project-jdk-name="11" project-jdk-type="JavaSDK" />
```
Second time I am seeing this file change by a different person. I wonder if everyone in the team should standardize on their JDK or simply remove this file from being tracked by git? Otherwise this file may be continually changed by everyone's unique environment.

kevlam2706 replied:
agreed. .gitignore time?

georgetzavelas commented on src/main/java/frc/robot/commands/DrivetrainPIDTurnDelta.java:
```diff
@@ -0,0 +1,67 @@
+
+    //get the drivebase and pigeon
+    private final DriveBase drivebase;
+    private PigeonIMU _pidgey = Pigeon.getPigeon();
```
Nitpicks so choose to ignore if you want: To be consistent I would initialize these in the constructor. Also for consistency I would remove the `_` from the variable name unless you are using it to signify something different about this variable vs the others?

kevlam2706 replied:
Or per my previous comment I think there is a case for having the Pigeon instantiated in the DriveBase and then this command would just do a getInstance of it.

georgetzavelas commented on src/main/java/frc/robot/config/Config.java:
```diff
@@ -63,11 +65,16 @@ private Config() {
 
+    static NetworkTable constantsTable = NetworkTableInstance.getDefault().getTable("constants");
+
+    public static FluidConstant<Double> DRIVETRAIN_P = new FluidConstant<>("DrivetrainP", 0.01d)
```
😻

georgetzavelas commented on src/main/java/frc/robot/sensors/Pigeon.java:
```diff
@@ -0,0 +1,27 @@
+
+    static {
+        //get talon
+        _pidgey = new PigeonIMU(DriveBase.getInstance().leftFrontTalon);
```
Have you thought about using a Singleton pattern here instead of initializing this statically? I would recommend a static class for independent calculations, e.g. math routines on passed in variables, but when accessing another instance of a shared resource I think a Singleton has less chance for subtle bugs, e.g. calling this static class from another static variable, since it controls access to the variable through a single instance of an object. Not that this is bad/wrong, just food for thought.

kevlam2706 replied:
Agreed. Let's get the architecture right from early on so that we start on the right foot and everyone can emulate good practices. The 'double instantiation of a single object' issue is a bug we have seen many many times over the years.

Again I think there's a case for the Pigeon to be managed in the DriveBase. I don't think I like the pattern of having a class (robot.sensors.Pigeon) that is this strongly linked to the DriveBase and in particular the internal specifics of the DriveBase (the fact that it has a left front Talon, this should really be private to the DriveBase. One day we might have a Falcon or NEO based DriveBase :) )

Please consider my suggestion of moving the management of the Pigeon to within the drivebase subsystem.

georgetzavelas commented on src/main/java/frc/robot/sensors/Pigeon.java:
```diff
@@ -0,0 +1,27 @@
+
+    public static double getCurrentAngle(){
+        //Gets the current angle
+        PigeonIMU.FusionStatus fusionStatus = new PigeonIMU.FusionStatus();
```
Should this be `_pidgey` instead of creating a new object?

kevlam2706 replied:
Agreed. Let's be very careful about instantiating multiple objects that tie back to a single piece of physical hardware. At worst, there will be a robot code crash (this used to happen to us a lot when we were allocating IOs on the RoboRio). At best this will lead to subtle bugs because one instance will tell the hardware to do X or keep track of Y and another instance will have no such knowledge and your code will get nulls or bad data or worse.

nrgill28 replied:
No this code is fine. The Pigeon library is very C-style where you instantiate a class and pass it into a function which modifies the class instead of returning a result

kevlam2706 replied:
OK, as long as that is clear to everyone. 

This is a general comment but remember that we are writing this code not just for this year but potentially for future years as well, including for rookie students. For that reason I would prioritize things like consistence, readability, clarity, and simplicity over clever tricks and special exceptions.

georgetzavelas reviewed (commented):
Amazing work! Some comments to consider

kevlam2706 commented on src/main/java/frc/robot/commands/DrivetrainPIDTurnDelta.java:
```diff
@@ -0,0 +1,67 @@
+        currentAngle = Pigeon.getCurrentAngle();
+
+        //Do PD
+        turnThrottle = (targetAngle - currentAngle) * pGain.get() - (currentAngularRate) * dGain.get();
```
Huh. Manual PID?

kevlam2706 commented on src/main/java/frc/robot/sensors/Pigeon.java:
```diff
@@ -0,0 +1,27 @@
+        _pidgey = new PigeonIMU(DriveBase.getInstance().leftFrontTalon);
+
+        //setup gyro angle finder
+        _pidgey.setFusedHeading(0.0, 30);
```
The side effect of running this command is that it will reset the pigeon heading.

That's OK if this is the convention we set, but everyone needs to be aware of it, e.g. the other students doing autonomous and vision commands.

The other way to do it is let the Pigeon keep track of its own heading and everyone uses their own variables to offset to what we want.

kevlam2706 reviewed (approved):
approved but please consider my request for changes to architecture. let's get this right before we let other students loose on it.

Lockscout committed:
Moved Pigeon to Drivebase

Moved the pigeon to drivebase

kevlam2706 reviewed (approved):
cool

Lockscout committed:
Added an argument to set drive forward speed

You can now set a speed the robot can drive forward at, from -1 to 1. If you set the delta to 0, it will now drive forward and keep straight even when pushed.

nrgill28 commented on src/main/java/frc/robot/commands/DrivetrainPIDTurnDelta.java:
```diff
@@ -64,6 +75,6 @@ public void execute() {
 
         //Run motors according to the output of PD
-        drivebase.tankDrive(-turnThrottle, turnThrottle, false);
+        drivebase.tankDrive(-turnThrottle + forwardSpeed, turnThrottle + forwardSpeed, false);
```
This might be an odd case, but since I think tankDrive clamps it's input to `0 <= x <= 1` you'll run in to some issues when `turnThrottle + forwardSpeed > 1` since the clamp will chop off part of your speed difference.  
I imagine it will only affect how fast you turn though.  

Also I know `A + B` is the same as `B + A` but I think it would look better with the variables swapped around 😛

Lockscout replied:
I think that it would be a rare case, since it would only matter a lot if we are going at max velocity, which I feel like we wouldn't do. I also feel like when used with vision they wouldn't want the robot moving at max speed since that would cause motion blur.

Lockscout committed:
Deleted .idea folder

Lockscout committed:
Comments and resolved a merge conflict

Commit:
Merge branch 'master' into DriveTrainPID
